WalletWasabi.Tests/UnitTests/BlockchainAnalysis/PubKeyReuseAnonScoreTests.cs

```diff
@@ -0,0 +1,154 @@
+				new[] { (Money.Coins(1m), HdPubKey.DefaultHighAnonymitySet, BitcoinFactory.CreateHdPubKey(km)) });
+
+			analyser.Analyze(tx);
+
+			Assert.All(tx.WalletInputs, x => Assert.Equal(100, x.HdPubKey.AnonymitySet));
+			Assert.Equal(100, tx.WalletOutputs.First().HdPubKey.AnonymitySet);
+		}
+
+		[Fact]
+		public void InputSideAddressReuseHaveNoConsolidationPunishmentInCoinJoin()
+		{
+			var analyser = ServiceFactory.CreateBlockchainAnalyzer();
+			var km = ServiceFactory.CreateKeyManager();
+			var key = BitcoinFactory.CreateHdPubKey(km);
+			var tx = BitcoinFactory.CreateSmartTransaction(
+				9,
+				Enumerable.Repeat(Money.Coins(1m), 9),
+				new[] { (Money.Coins(1.1m), 100, key), (Money.Coins(1.2m), 100, key), (Money.Coins(1.3m), 100, key), (Money.Coins(1.4m), 100, key) },
+				new[] { (Money.Coins(1m), HdPubKey.DefaultHighAnonymitySet, BitcoinFactory.CreateHdPubKey(km)) });
+
+			analyser.Analyze(tx);
+
+			Assert.All(tx.WalletInputs, x => Assert.Equal(100, x.HdPubKey.AnonymitySet));
+			Assert.Equal(109, tx.WalletOutputs.First().HdPubKey.AnonymitySet);
+		}
+
+		[Fact]
+		public void InputOutputSideAddress()
+		{
+			// If there's reuse in input and output side, then output side didn't gain, nor lose anonymity.
+			var analyser = ServiceFactory.CreateBlockchainAnalyzer();
+			var key = BitcoinFactory.CreateHdPubKey(ServiceFactory.CreateKeyManager());
+			var tx = BitcoinFactory.CreateSmartTransaction(
+				9,
+				Enumerable.Repeat(Money.Coins(1m), 9),
+				new[] { (Money.Coins(1.1m), 100, key) },
+				new[] { (Money.Coins(1m), HdPubKey.DefaultHighAnonymitySet, key) });
+
+			analyser.Analyze(tx);
+
+			Assert.All(tx.WalletInputs, x => Assert.Equal(100, x.HdPubKey.AnonymitySet));
+			Assert.Equal(100, tx.WalletOutputs.First().HdPubKey.AnonymitySet);
+		}
+
+		[Fact]
+		public void OutputSideAddressReusePunished()
+		{
+			// If there's reuse in input and output side, then output side didn't gain, nor lose anonymity.
+			var analyser = ServiceFactory.CreateBlockchainAnalyzer();
+			var km = ServiceFactory.CreateKeyManager();
+			var key = BitcoinFactory.CreateHdPubKey(km);
+			var tx = BitcoinFactory.CreateSmartTransaction(
+				9,
+				Enumerable.Repeat(Money.Coins(1m), 9).Concat(Enumerable.Repeat(Money.Coins(2m), 7)),
+				new[] { (Money.Coins(1.1m), 100, BitcoinFactory.CreateHdPubKey(km)) },
+				new[] { (Money.Coins(1m), HdPubKey.DefaultHighAnonymitySet, key), (Money.Coins(2m), HdPubKey.DefaultHighAnonymitySet, key) });
+
+			analyser.Analyze(tx);
+
+			Assert.All(tx.WalletInputs, x => Assert.Equal(100, x.HdPubKey.AnonymitySet));
+
+			// Normally all levels should have 109 and 106 anonsets, but they're consolidated and punished.
+			Assert.All(tx.WalletOutputs.Select(x => x.HdPubKey.AnonymitySet), x => Assert.True(x < 106));
+		}
+
+		[Fact]
+		public void OutputSideAddressReuseDoesntPunishedMoreThanInheritance()
```

Comment:
similar to the above failing test, what if all these outputs are using a previously used key? i think in that case they should lose anon

Reply:
Yup, they do. Look at `AddressReusePunishment` test.

I do a transaction with 100 input anonset and a would be 10 anonset gain for the output, which would end up having 109 anonset finally.

The trick is that, then I make sure there's an address reuse with a 30 anonset key: `reuse.AnonymitySet = 30;`.

Then I assert that the reuse's anonset isn't 109, isn't 30, but that it's less than 30, because of the intersection punishment.